src/main.ml

```diff
@@ -79,25 +79,42 @@ let handle_connection fd tls_role =
     )
 
 (* TODO use the version from nbd repository *)
-let init_tls_get_server_ctx ~certfile ~ciphersuites no_tls =
-  if no_tls then None
-  else (
-    let certfile = require_str "certfile" certfile in
-    let ciphersuites = require_str "ciphersuites" ciphersuites in
-    Some (Nbd_lwt_unix.TlsServer
-      (Nbd_lwt_unix.init_tls_get_ctx ~certfile ~ciphersuites)
-    )
+let init_tls_get_server_ctx ~certfile ~ciphersuites =
+  let certfile = require_str "certfile" certfile in
+  let ciphersuites = require_str "ciphersuites" ciphersuites in
+  Some (Nbd_lwt_unix.TlsServer
+    (Nbd_lwt_unix.init_tls_get_ctx ~certfile ~ciphersuites)
   )
 
-let main port certfile ciphersuites no_tls =
+let xapi_says_use_tls () =
+  let refuse msg = (
+    Lwt_log.notice msg >>=
+    (fun () -> Lwt.fail_with msg)
+  ) in
+  let ask_xapi rpc session_id =
+    Xen_api.Network.get_all_records ~rpc ~session_id >>=
+    fun all_nets ->
+    let all_porpoises = List.map (fun (_str, net) -> net.API.network_purpose) all_nets |>
+    List.flatten in
+    let tls = List.mem `nbd all_porpoises in
+    let no_tls = List.mem `insecure_nbd all_porpoises in
+    match tls, no_tls with
+    | true, true -> refuse "Contradictory XenServer configuration: nbd and insecure_nbd network purposes!"
+    | true, false -> Lwt.return true
+    | false, true -> Lwt.return false
+    | false, false -> refuse "No network has purpose nbd or insecure_nbd."
+  in
+  Cleanup.Local_xapi_session.with_session ask_xapi
+
+let main port certfile ciphersuites =
   let t () =
-    Lwt_log.notice_f "Starting xapi-nbd: port = '%d'; certfile = '%s'; ciphersuites = '%s' no_tls = '%b'" port certfile ciphersuites no_tls >>= fun () ->
+    Lwt_log.notice_f "Starting xapi-nbd: port = '%d'; certfile = '%s'; ciphersuites = '%s'" port certfile ciphersuites >>= fun () ->
     (* We keep a persistent record of the VBDs that we've created but haven't
        yet cleaned up. At startup we go through this list in case some VBDs
        got leaked after the previous run due to a crash and clean them up. *)
     Cleanup.Persistent.cleanup () >>= fun () ->
     Lwt_log.notice "Initialising TLS" >>= fun () ->
-    let tls_role = init_tls_get_server_ctx ~certfile ~ciphersuites no_tls in
+    let tls_server_role = init_tls_get_server_ctx ~certfile ~ciphersuites in
```

Comment:
We could try to make this a lazy value so that we don't init TLS when we don't need it. Although that would make testing a bit harder. So let's not do it for now.